doc/compiler/debug/Loggers.md

```diff
@@ -0,0 +1,176 @@
+underlying Logger will never be exceeded. Instead, it makes a reasonable effort
+to keep the logging output within a certain character threshold. This is
+identical behaviour to the original Testarossa circular output throttle that
+existed before Loggers.
+
+### MemoryBufferLogger
+
+A MemoryBufferLogger logs its output to a provided fixed-size memory buffer.
+Logging output is always guaranteed to fall within the bounds of the buffer,
+with truncation occurring as necessary. The memory buffer will be
+'\0'-terminated and the memory buffer must be sized to accommodate the
+termination character.
+
+### AssertingLogger
+
+An AssertingLogger is a Logger where each API output function fatal asserts if
+it is called. It is useful in test modes to catch instances of unguarded Logger
+output.
+
+## Using Loggers
+
+### Logger Availability
+
+To ensure versatility, all diagnostic functions should accept a Logger
+parameter for the logging destination. By convention, this should be passed as
+the first parameter. There may be circumstances where a Logger is not passed,
+but only in situations where the logging destination can be obviously derived
+from the context.
+
+Stated differently, diagnostic functions should not make an assumption that
+their output will go to the Logger found on the `TR::Compilation` object. This
+makes it difficult to support creative debugging techniques (such as dumping a
+data structure to an alternate file or simply stdio) because the destination is
+fixed.
+
+Nevertheless, a valid Logger will always be available on the `TR::Compilation`
+object to catch diagnostic output. The default Logger is the `TR::NullLogger`
+which simply eats anything passed to it. This differs from the historical
+Testarossa design where a log file is only created when specified by the user
+on the command-line, leading to defensive code that repeatedly confirmed the
+existence of a log file throughout the code. Such checks no longer apply,
+however, all diagnastic output should be guarded with a trace option of some
+sort. The default Logger is replaced with a more appropriate Logger depending
+on the diagnostic circumstances. Downstream projects may use a
+`TR::AssertingLogger` as part of their testing to catch unguarded use of
+logging functions.
+
+### Guarding Diagnostic Logging
+
+All diagnostic output should be guarded with a trace option that controls its
+output. Controlling diagnostic output by checking for the presence of a log
+file is a deprecated anti-pattern.
+
+A common pattern for guarding a single output function is the following:
+```
+   if (myTraceOption)
+      {
+      log->printf("My output: %d\n", myInt);
+      }
+```
+To improve code readability and density, convenience macros have been introduced
+for the "print" family of functions that should be used in lieu of the pattern
+above. For example, the code above is equivalent to:
+```
+   trprintf(myTraceOption, log, "My output: %d\n", myInt);
+```
```

Comment:
May I suggest renaming `log` to `logger` in these examples to make it clear that those are Logger objects, rather than log files?

Reply:
No, I definitely wouldn't want `logger` changed to `log` throughout, and I think it would be preferable to have the documentation match expected usage, so I will withdraw this comment.